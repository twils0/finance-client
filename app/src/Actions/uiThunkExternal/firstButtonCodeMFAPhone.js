import { codeTypeNames } from '../../Constants/dataConstantsAuth';
import { errorMessages } from '../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { fieldNames } from '../../Constants/dataConstantsAccount';
import requestLoginMFA from '../dataThunkAuth/requestLoginMFA';
import { setCurrentForm, setInputValueError } from '../uiActionsExternal';

const firstButtonCodePhone = () => async (dispatch, getState) => {
  let state = getState();
  const { forms } = state.ui.external;

  const codeCodeMFAPhone = forms[formNames.CODE_MFA_PHONE].inputs[inputNames[formNames.CODE_MFA_PHONE].CODE];

  if (!codeCodeMFAPhone.value) {
    codeCodeMFAPhone.errorMessage = errorMessages.NO_CONFIRMATION_CODE_PHONE;
  }

  if (codeCodeMFAPhone.errorMessage) {
    dispatch(setInputValueError(codeCodeMFAPhone));
  } else {
    try {
      await dispatch(requestLoginMFA({ code: codeCodeMFAPhone.value }));

      state = getState();
      const { codeTypes } = state.data.auth;
      const emailAdditional = state.data.account.fields[fieldNames.EMAIL_ADDITIONAL].value;

      if (
        codeTypes[codeTypeNames.VERIFY_EMAIL].needed
        || (codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed && emailAdditional)
      ) {
        dispatch(setCurrentForm({ current: formNames.CODE_VERIFY_EMAIL }));
      } else {
        dispatch(setCurrentForm({ current: formNames.DEVICE }));
      }

      if (codeCodeMFAPhone.value || codeCodeMFAPhone.errorMessage) {
        codeCodeMFAPhone.value = '';
        codeCodeMFAPhone.errorMessage = '';

        dispatch(setInputValueError(codeCodeMFAPhone));
      }
    } catch (error) {
      switch (error.code) {
        case 'LimitExceededException': {
          codeCodeMFAPhone.errorMessage = errorMessages.LIMIT_EXCEEDED;
          dispatch(setInputValueError(codeCodeMFAPhone));
          break;
        }
        case 'ExpiredCodeException': {
          codeCodeMFAPhone.errorMessage = errorMessages.EXPIRED_CONFIRMATION_CODE;
          dispatch(setInputValueError(codeCodeMFAPhone));
          break;
        }
        case 'CodeMismatchException': {
          codeCodeMFAPhone.errorMessage = errorMessages.INVALID_CONFIRMATION_CODE;
          dispatch(setInputValueError(codeCodeMFAPhone));
          break;
        }
        default: {
          codeCodeMFAPhone.errorMessage = errorMessages.INTERNAL_ERROR;
          dispatch(setInputValueError(codeCodeMFAPhone));
          break;
        }
      }
    }
  }

  return null;
};

export default firstButtonCodePhone;
