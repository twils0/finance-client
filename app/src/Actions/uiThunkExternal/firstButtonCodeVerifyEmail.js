import { requestStatusTypes, pathNames } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { setAuthenticated } from '../dataActionsAuth';
import requestVerifyEmailLink from '../dataThunkAuth/requestVerifyEmailLink';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import { setCurrentForm, setInputValueError } from '../uiActionsExternal';

const firstButtonCodeEmail = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'match')) {
    throw new Error(`Please enter a value for the 'match' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    const state = getState();
    let login = false;
    const { verificationId } = payload.match.params;
    const { status } = state.data.auth;

    try {
      if (verificationId) {
        let { user } = state.data.aws;

        if (!user) {
          ({ user } = await dispatch(requestAWSUser()));
        }

        await dispatch(requestVerifyEmailLink({ verificationId }));

        if (user) {
          login = true;
        } else {
          payload.history.replace(pathNames.LOGIN);
        }
      } else if (status[statusNames.LOGIN_MFA].status === requestStatusTypes.SUCCESS) {
        dispatch(setCurrentForm({ current: formNames.DEVICE }));
      } else {
        login = true;
      }

      if (login) {
        const { forms } = state.ui.external;
        const loginEmail = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
        const loginPassword = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];
        const currentTime = new Date();

        currentTime.setMinutes(currentTime.getMinutes() + 30);
        window.sessionStorage.setSecurity('sessionTime', currentTime);

        dispatch(setAuthenticated({ authenticated: true }));

        if (loginEmail.value || loginEmail.errorMessage) {
          loginEmail.value = '';
          loginEmail.errorMessage = '';

          dispatch(setInputValueError(loginEmail));
        }

        if (loginPassword.value || loginPassword.errorMessage) {
          loginPassword.value = '';
          loginPassword.errorMessage = '';

          dispatch(setInputValueError(loginPassword));
        }
      }
    } catch (errorCatch) {
      payload.history.replace(pathNames.LOGIN);
    }

    return null;
  };
};

export default firstButtonCodeEmail;
