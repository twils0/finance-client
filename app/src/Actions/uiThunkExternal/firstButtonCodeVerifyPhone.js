import { pathNames, requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames, codeTypeNames } from '../../Constants/dataConstantsAuth';
import { fieldNames } from '../../Constants/dataConstantsAccount';
import { errorMessages } from '../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { setAuthenticated } from '../dataActionsAuth';
import { setCurrentForm, setInputValueError } from '../uiActionsExternal';
import requestVerifyFieldConfirm from '../dataThunkAuth/requestVerifyFieldConfirm';

const firstButtonCodePhone = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'history')) {
    throw new Error(`Please enter a value for the 'history' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.external;

    const codeCodeVerifyPhone = forms[formNames.CODE_VERIFY_PHONE].inputs[inputNames[formNames.CODE_VERIFY_PHONE].CODE];

    if (!codeCodeVerifyPhone.value) {
      codeCodeVerifyPhone.errorMessage = errorMessages.NO_VERIFICATION_CODE_PHONE;
    }

    if (codeCodeVerifyPhone.errorMessage) {
      dispatch(setInputValueError(codeCodeVerifyPhone));
    } else {
      const { status, codeTypes } = state.data.auth;
      const emailAdditional = state.data.account.fields[fieldNames.EMAIL_ADDITIONAL].value;

      try {
        await dispatch(
          requestVerifyFieldConfirm({
            field: 'phone_number',
            code: codeCodeVerifyPhone.value,
          }),
        );

        if (
          codeTypes[codeTypeNames.VERIFY_EMAIL].needed
          || (codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed && emailAdditional)
        ) {
          dispatch(setCurrentForm({ current: formNames.CODE_VERIFY_EMAIL }));
        } else if (status[statusNames.LOGIN_MFA].status === requestStatusTypes.SUCCESS) {
          dispatch(setCurrentForm({ current: formNames.DEVICE }));
        } else {
          const loginEmail = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
          const loginPassword = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];

          const currentTime = new Date();

          currentTime.setMinutes(currentTime.getMinutes() + 30);
          window.sessionStorage.setSecurity('sessionTime', currentTime);

          dispatch(setAuthenticated({ authenticated: true }));

          if (loginEmail.value || loginEmail.errorMessage) {
            loginEmail.value = '';
            loginEmail.errorMessage = '';

            dispatch(setInputValueError(loginEmail));
          }

          if (loginPassword.value || loginPassword.errorMessage) {
            loginPassword.value = '';
            loginPassword.errorMessage = '';

            dispatch(setInputValueError(loginPassword));
          }
        }

        if (codeCodeVerifyPhone.value || codeCodeVerifyPhone.errorMessage) {
          codeCodeVerifyPhone.value = '';
          codeCodeVerifyPhone.errorMessage = '';

          dispatch(setInputValueError(codeCodeVerifyPhone));
        }
      } catch (error) {
        if (typeof error === 'string' && error === 'Username cannot be empty') {
          payload.history.replace(pathNames.LOGIN);

          if (codeCodeVerifyPhone.value || codeCodeVerifyPhone.errorMessage) {
            codeCodeVerifyPhone.value = '';
            codeCodeVerifyPhone.errorMessage = '';

            dispatch(setInputValueError(codeCodeVerifyPhone));
          }
        }

        if (typeof error === 'object' && error && error.code) {
          switch (error.code) {
            case 'LimitExceededException': {
              codeCodeVerifyPhone.errorMessage = errorMessages.LIMIT_EXCEEDED;
              dispatch(setInputValueError(codeCodeVerifyPhone));
              break;
            }
            case 'ExpiredCodeException': {
              codeCodeVerifyPhone.errorMessage = errorMessages.EXPIRED_VERIFICATION_CODE;
              dispatch(setInputValueError(codeCodeVerifyPhone));
              break;
            }
            case 'CodeMismatchException': {
              codeCodeVerifyPhone.errorMessage = errorMessages.INVALID_VERIFICATION_CODE;
              dispatch(setInputValueError(codeCodeVerifyPhone));
              break;
            }
            default: {
              codeCodeVerifyPhone.errorMessage = errorMessages.INTERNAL_ERROR;
              dispatch(setInputValueError(codeCodeVerifyPhone));
              break;
            }
          }
        }
      }
    }

    return null;
  };
};

export default firstButtonCodePhone;
