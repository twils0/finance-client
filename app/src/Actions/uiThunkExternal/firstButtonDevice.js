import { pathNames } from '../../Constants/universalConstants';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { setAuthenticated } from '../dataActionsAuth';
import requestSignOutOtherDevices from '../dataThunkAuth/requestSignOutOtherDevices';
import { setInputValueError } from '../uiActionsExternal';

const firstButtonDevice = payload => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'history')) {
    throw new Error(`Please enter a value for the 'history' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.external;

    const loginEmail = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
    const loginPassword = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];

    const signOutPayload = {
      email: loginEmail.value,
      password: loginPassword.value,
      remembered: true,
    };

    try {
      await dispatch(requestSignOutOtherDevices(signOutPayload));

      dispatch(setAuthenticated({ authenticated: true }));

      if (loginEmail.value || loginEmail.errorMessage) {
        loginEmail.value = '';
        loginEmail.errorMessage = '';

        dispatch(setInputValueError(loginEmail));
      }

      if (loginPassword.value || loginPassword.errorMessage) {
        loginPassword.value = '';
        loginPassword.errorMessage = '';

        dispatch(setInputValueError(loginPassword));
      }
    } catch (error) {
      payload.history.replace(pathNames.LOGIN);
    }

    return null;
  };
};

export default firstButtonDevice;
