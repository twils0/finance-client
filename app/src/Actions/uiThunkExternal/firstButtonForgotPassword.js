import isEmail from 'validator/lib/isEmail';

import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { errorMessages } from '../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { setAuthStatus } from '../dataActionsAuth';
import requestForgotPassword from '../dataThunkAuth/requestForgotPassword';
import { setCurrentForm, setInputValueError } from '../uiActionsExternal';

const firstButtonForgotPassword = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'history')) {
    throw new Error(`Please enter a value for the 'history' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    const state = getState();
    const { forms } = state.ui.external;

    const loginEmail = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
    const loginPassword = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];
    const forgotPasswordEmail = forms[formNames.FORGOT_PASSWORD].inputs[inputNames[formNames.FORGOT_PASSWORD].EMAIL];

    if (!forgotPasswordEmail.value) {
      forgotPasswordEmail.errorMessage = errorMessages.NO_EMAIL;
    } else if (!isEmail(forgotPasswordEmail.value)) {
      forgotPasswordEmail.errorMessage = errorMessages.INVALID_EMAIL;
    }

    if (forgotPasswordEmail.errorMessage) {
      dispatch(setInputValueError(forgotPasswordEmail));
    } else {
      try {
        await dispatch(requestForgotPassword({ email: forgotPasswordEmail.value }));

        dispatch(setCurrentForm({ current: formNames.RESET_PASSWORD }));

        if (loginEmail.value || loginEmail.errorMessage) {
          loginEmail.value = '';
          loginEmail.errorMessage = '';

          dispatch(setInputValueError(loginEmail));
        }

        if (loginPassword.value || loginPassword.errorMessage) {
          loginPassword.value = '';
          loginPassword.errorMessage = '';

          dispatch(setInputValueError(loginPassword));
        }

        dispatch(
          setAuthStatus({
            id: statusNames.FORGOT_PASSWORD,
            status: requestStatusTypes.IDLE,
          }),
        );
      } catch (error) {
        switch (error.code) {
          case 'LimitExceededException': {
            forgotPasswordEmail.errorMessage = errorMessages.LIMIT_EXCEEDED;
            dispatch(setInputValueError(forgotPasswordEmail));
            break;
          }
          case 'UserNotFoundException': {
            forgotPasswordEmail.errorMessage = errorMessages.EMAIL_NOT_FOUND;
            dispatch(setInputValueError(forgotPasswordEmail));
            break;
          }
          default: {
            forgotPasswordEmail.errorMessage = errorMessages.INTERNAL_ERROR;
            dispatch(setInputValueError(forgotPasswordEmail));
            break;
          }
        }
      }
    }

    return null;
  };
};

export default firstButtonForgotPassword;
