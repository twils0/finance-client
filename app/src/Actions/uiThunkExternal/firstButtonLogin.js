import isEmail from 'validator/lib/isEmail';

import { pathNames } from '../../Constants/universalConstants';
import { codeTypeNames } from '../../Constants/dataConstantsAuth';
import { errorMessages } from '../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../Constants/uiConstantsExternal';
import { fieldNames } from '../../Constants/dataConstantsAccount';
import requestLogin from '../dataThunkAuth/requestLogin';
import requestSignUpResendPhone from '../dataThunkAuth/requestSignUpResendPhone';
import { setCurrentForm, setInputValueError } from '../uiActionsExternal';

const firstButtonLogin = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'history')) {
    throw new Error(`Please enter a value for the 'history' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    let state = getState();
    const { forms } = state.ui.external;

    const loginEmail = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].EMAIL];
    const loginPassword = forms[formNames.LOGIN].inputs[inputNames[formNames.LOGIN].PASSWORD];

    if (!loginEmail.value) {
      loginEmail.errorMessage = errorMessages.NO_EMAIL;
    } else if (!isEmail(loginEmail.value)) {
      loginEmail.errorMessage = errorMessages.INVALID_EMAIL;
    }

    if (!loginPassword.value) {
      loginPassword.errorMessage = errorMessages.NO_PASSWORD;
    }

    if (loginEmail.errorMessage || loginPassword.errorMessage) {
      dispatch(setInputValueError(loginEmail));
      dispatch(setInputValueError(loginPassword));
    } else {
      const loginPayload = {
        email: loginEmail.value,
        password: loginPassword.value,
      };

      try {
        const mfa = await dispatch(requestLogin(loginPayload));

        state = getState();
        const { codeTypes } = state.data.auth;
        const emailAdditional = state.data.account.fields[fieldNames.EMAIL_ADDITIONAL].value;

        if (typeof mfa === 'boolean' && mfa) {
          payload.history.replace(pathNames.CODE);
        } else if (codeTypes[codeTypeNames.VERIFY_PHONE].needed) {
          payload.history.replace(pathNames.CODE);
          dispatch(setCurrentForm({ current: formNames.CODE_VERIFY_PHONE }));
        } else if (
          codeTypes[codeTypeNames.VERIFY_EMAIL].needed
          || (codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed && emailAdditional)
        ) {
          payload.history.replace(pathNames.CODE);
          dispatch(setCurrentForm({ current: formNames.CODE_VERIFY_EMAIL }));
        }
      } catch (error) {
        switch (error.code) {
          case 'LimitExceededException': {
            loginPassword.errorMessage = errorMessages.LIMIT_EXCEEDED;
            dispatch(setInputValueError(loginPassword));
            break;
          }
          case 'UserNotFoundException': {
            loginEmail.errorMessage = errorMessages.EMAIL_NOT_FOUND;
            dispatch(setInputValueError(loginEmail));
            break;
          }
          case 'NotAuthorizedException': {
            if (error.message === 'Incorrect username or password.') {
              loginPassword.errorMessage = errorMessages.PASSWORD_NOT_FOUND;
              dispatch(setInputValueError(loginPassword));
            } else {
              loginPassword.errorMessage = errorMessages.INTERNAL_ERROR;
              dispatch(setInputValueError(loginPassword));
            }
            break;
          }
          case 'UserNotConfirmedException': {
            dispatch(requestSignUpResendPhone({ email: loginEmail.value }));

            payload.history.replace(pathNames.CODE);
            dispatch(setCurrentForm({ current: formNames.CODE_VERIFY_PHONE }));
            break;
          }
          default: {
            loginPassword.errorMessage = errorMessages.INTERNAL_ERROR;
            dispatch(setInputValueError(loginPassword));
            break;
          }
        }
      }
    }

    return null;
  };
};

export default firstButtonLogin;
