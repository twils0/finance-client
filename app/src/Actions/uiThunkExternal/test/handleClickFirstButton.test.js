import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { initialState as initialStateUIExternal } from '../../../Reducers/uiReducersExternal';
import {
  actionTypes as actionTypesUIExternal,
  formNames,
} from '../../../Constants/uiConstantsExternal';
import handleClickFirstButton from '../handleClickFirstButton';
import firstButtonResetPassword from '../firstButtonResetPassword';
import firstButtonForgotPassword from '../firstButtonForgotPassword';
import firstButtonLogin from '../firstButtonLogin';
import firstButtonSignUp from '../firstButtonSignUp';
import firstButtonCodeMFAPhone from '../firstButtonCodeMFAPhone';
import firstButtonCodeVerifyPhone from '../firstButtonCodeVerifyPhone';
import firstButtonCodeVerifyEmail from '../firstButtonCodeVerifyEmail';
import firstButtonDevice from '../firstButtonDevice';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

jest.mock('../firstButtonResetPassword', () => jest.fn());
jest.mock('../firstButtonForgotPassword', () => jest.fn());
jest.mock('../firstButtonLogin', () => jest.fn());
jest.mock('../firstButtonSignUp', () => jest.fn());
jest.mock('../firstButtonCodeMFAPhone', () => jest.fn());
jest.mock('../firstButtonCodeVerifyPhone', () => jest.fn());
jest.mock('../firstButtonCodeVerifyEmail', () => jest.fn());
jest.mock('../firstButtonDevice', () => jest.fn());

const history = { test: 'testHistory' };
const match = { test: 'testMatch' };
const stripeElement = { test: 'testStripeElement' };
const resetStripeElement = { test: 'testReset' };
const handlerPayload = {
  history,
  match,
  stripeElement,
  resetStripeElement,
};

describe('uiThunkExternal', () => {
  describe('handleClickFirstButton', () => {
    afterEach(() => {
      firstButtonResetPassword.mockReset();
      firstButtonForgotPassword.mockReset();
      firstButtonLogin.mockReset();
      firstButtonSignUp.mockReset();
      firstButtonCodeMFAPhone.mockReset();
      firstButtonCodeVerifyPhone.mockReset();
      firstButtonCodeVerifyEmail.mockReset();
      firstButtonDevice.mockReset();
    });

    it("fails and throws an error when missing 'history' key in payload", async () => {
      const store = mockStore({
        ui: { external: initialStateUIExternal },
      });

      const emptyPayload = {};

      try {
        await store.dispatch(handleClickFirstButton(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }
    });

    it("fails and throws an error when missing 'history' key in payload", async () => {
      const store = mockStore({
        ui: { external: initialStateUIExternal },
      });

      const emptyPayload = { history };

      try {
        await store.dispatch(handleClickFirstButton(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }
    });

    it("fails and throws an error when missing 'stripeElement' key in payload", async () => {
      const store = mockStore({
        ui: { external: initialStateUIExternal },
      });

      const wrongPayload = { history, match };

      try {
        await store.dispatch(handleClickFirstButton(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }
    });

    it("fails and throws an error when missing 'resetStripeElement' key in payload", async () => {
      const store = mockStore({
        ui: { external: initialStateUIExternal },
      });

      const wrongPayload = { history, match, stripeElement };

      try {
        await store.dispatch(handleClickFirstButton(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }
    });

    it('calls nothing when forms current is wrong', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = 'testForm';

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
    });

    it('calls firstButtonResetPassword when form current is RESET_PASSWORD', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.RESET_PASSWORD;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonResetPassword.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonResetPassword).toBeCalled();
    });

    it('calls firstButtonForgotPassword when form current is FORGOT_PASSWORD', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.FORGOT_PASSWORD;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonForgotPassword.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonForgotPassword).toBeCalledWith({ history });
    });

    it('calls firstButtonLogin when form current is LOGIN', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.LOGIN;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonLogin.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonLogin).toBeCalledWith({ history });
    });

    it('calls firstButtonLogin when form current is CARD', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.CARD;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      const expectedActions = [
        {
          type: actionTypesUIExternal.SET_CURRENT_FORM,
          payload: {
            current: formNames.INFO,
          },
        },
      ];

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
    });

    it('calls firstButtonLogin when form current is INFO', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.INFO;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      const expectedActions = [
        {
          type: actionTypesUIExternal.SET_CURRENT_FORM,
          payload: {
            current: formNames.SIGN_UP,
          },
        },
      ];

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(null);
    });

    it('calls firstButtonSignUp when form current is SIGN_UP', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.SIGN_UP;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonSignUp.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonSignUp).toBeCalledWith({ history, stripeElement, resetStripeElement });
    });

    it('calls firstButtonCodeMFAPhone when form current is CODE_MFA_PHONE', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.CODE_MFA_PHONE;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonCodeMFAPhone.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonCodeMFAPhone).toBeCalled();
    });

    it('calls firstButtonCodeVerifyPhone when form current is CODE_VERIFY_PHONE', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.CODE_VERIFY_PHONE;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonCodeVerifyPhone.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonCodeVerifyPhone).toBeCalledWith({ history });
    });

    it('calls firstButtonCodeVerifyEmail when form current is CODE_VERIFY_EMAIL', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.CODE_VERIFY_EMAIL;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonCodeVerifyEmail.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonCodeVerifyEmail).toBeCalledWith({ history, match });
    });

    it('calls firstButtonDevice when form current is DEVICE', async () => {
      const stateBeforeUIExternal = JSON.parse(JSON.stringify(initialStateUIExternal));
      stateBeforeUIExternal.forms.current = formNames.DEVICE;

      const store = mockStore({ ui: { external: stateBeforeUIExternal } });

      firstButtonDevice.mockReturnValue(() => null);

      const result = store.dispatch(handleClickFirstButton(handlerPayload));

      expect(result).toEqual(null);
      expect(firstButtonDevice).toBeCalled();
    });
  });
});
