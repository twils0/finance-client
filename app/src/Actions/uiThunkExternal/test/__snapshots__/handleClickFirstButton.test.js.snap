// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uiThunkExternal handleClickFirstButton fails and throws an error when missing 'history' key in payload 1`] = `[Error: Please enter a value for the 'history' key - {}]`;

exports[`uiThunkExternal handleClickFirstButton fails and throws an error when missing 'history' key in payload 2`] = `[Error: Please enter a value for the 'match' key - {"history":{"test":"testHistory"}}]`;

exports[`uiThunkExternal handleClickFirstButton fails and throws an error when missing 'resetStripeElement' key in payload 1`] = `[Error: Please enter a value for the 'resetStripeElement' key - {"history":{"test":"testHistory"},"match":{"test":"testMatch"},"stripeElement":{"test":"testStripeElement"}}]`;

exports[`uiThunkExternal handleClickFirstButton fails and throws an error when missing 'stripeElement' key in payload 1`] = `[Error: Please enter a value for the 'stripeElement' key - {"history":{"test":"testHistory"},"match":{"test":"testMatch"}}]`;
